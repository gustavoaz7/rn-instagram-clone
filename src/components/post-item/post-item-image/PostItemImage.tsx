import React, { memo, useCallback, useMemo, useState } from 'react';
import styled, { useTheme } from 'styled-components/native';
import { View, Animated } from 'react-native';
import {
  TapGestureHandler,
  State,
  HandlerStateChangeEvent,
  TapGestureHandlerEventPayload,
  TapGestureHandlerProps,
} from 'react-native-gesture-handler';
import { SCREEN_WIDTH } from '../../../utils/dimensions';
import { TPostMedia } from '../../../types';
import { MentionTag } from '../../mention-tag';

type TPostItemImageProps = {
  media: TPostMedia;
  waitFor?: TapGestureHandlerProps['waitFor'];
};

export const PostItemImage = memo(
  ({ media, waitFor }: TPostItemImageProps): JSX.Element => {
    const theme = useTheme();
    const [showTags, setShowTags] = useState(false);
    const tagScale = useMemo(() => new Animated.Value(0), []);

    const onSingleTapMedia = useCallback(
      (e: HandlerStateChangeEvent<TapGestureHandlerEventPayload>) => {
        if (e.nativeEvent.state !== State.ACTIVE) return;
        const wasShowingTags = showTags;
        setShowTags(true);

        Animated.timing(tagScale, {
          toValue: showTags ? 0 : 1,
          duration: theme.animation.timingFast,
          useNativeDriver: true,
        }).start(({ finished }) => {
          if (finished) {
            setShowTags(!wasShowingTags);
          }
        });
      },
      [showTags, tagScale, theme.animation],
    );

    return (
      <TapGestureHandler
        onHandlerStateChange={onSingleTapMedia}
        waitFor={waitFor}
      >
        <View testID="PostItemImage">
          <Image source={{ uri: media.url }} />
          {showTags
            ? media.tappableObjects
                .filter(tappable => tappable.type === 'mention')
                .map(tappable => (
                  <AnimatedMentionTag
                    key={tappable.id}
                    style={{ transform: [{ scale: tagScale }] }}
                    {...{ top: tappable.y, left: tappable.x }}
                  >
                    <MentionTag {...tappable} />
                  </AnimatedMentionTag>
                ))
            : null}
        </View>
      </TapGestureHandler>
    );
  },
);

const Image = styled.Image`
  width: ${SCREEN_WIDTH}px;
  height: ${SCREEN_WIDTH}px;
`;

const AnimatedMentionTag = styled(Animated.View)<{ top: number; left: number }>`
  position: absolute;
  ${({ top, left }) => `
    top: ${top * 100}%;
    left: ${left * 100}%;
  `}
`;
