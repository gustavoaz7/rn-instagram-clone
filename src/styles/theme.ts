import type { Animated } from 'react-native';

const HEART_SPRING_SCALE_CONFIG: Animated.SpringAnimationConfig = {
  toValue: 1,
  useNativeDriver: true,
  velocity: 5,
  tension: 100,
  friction: 5,
};

export const theme = {
  spacing: {
    xs: '5px',
    s: '10px',
    m: '15px',
    l: '20px',
    xl: '25px',
  },
  color: {
    lightBlue: '#0095f6',
    royalBlue: '#405DE6',
    blue: '#5B51D8',
    purple: '#833AB4',
    darkPink: '#C13584',
    purpleRed: '#E1306C',
    red: '#FD1D1D',
    darkOrange: '#F56040',
    orange: '#F77737',
    yellow: '#FCAF45',
    lightYellow: '#FFDC80',
    white: '#FFFFFF',
    black: '#262626',
    gray: '#8e8e8e',
  },
  font: {
    size: {
      s: '12px',
      m: '14px',
      l: '16px',
      xl: '20px',
    },
  },
  animation: {
    timingFast: 150,
    timingBase: 300,
    timingSlow: 500,

    heart: {
      initialScale: 0.3,
      springConfig: HEART_SPRING_SCALE_CONFIG,
    },
  },
};

export type TTheme = typeof theme;
