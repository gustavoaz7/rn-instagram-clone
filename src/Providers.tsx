import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
