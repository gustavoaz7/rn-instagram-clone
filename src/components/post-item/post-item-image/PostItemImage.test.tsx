import React from 'react';
import { Image, Animated } from 'react-native';
import { render, act } from '@testing-library/react-native';
import { TapGestureHandler, State } from 'react-native-gesture-handler';
import { PostItemImage } from './PostItemImage';
import { Providers } from '../../../Providers';
import {
  generateMockPostMedia,
  generateMockTappableObject,
} from '../../../data';
import { theme } from '../../../styles/theme';
import {
  destroyTimeTravel,
  setupTimeTravel,
  timeTravel,
} from '../../../test/time-travel';

describe('components - PostItemImage', () => {
  const options = { wrapper: Providers };
  const media = generateMockPostMedia();

  it('renders ', () => {
    render(<PostItemImage media={media} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<PostItemImage media={media} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders image', () => {
    const { UNSAFE_getByType } = render(
      <PostItemImage media={media} />,
      options,
    );

    expect(UNSAFE_getByType(Image).props.source.uri).toEqual(media.url);
  });

  describe('when post has no mention', () => {
    it('does nothing on single-tap', async () => {
      const { UNSAFE_getByType, queryByTestId } = render(
        <PostItemImage media={media} />,
        options,
      );

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
      });

      expect(queryByTestId('MentionTag')).toBeNull();
    });
  });

  describe('when post has mentions', () => {
    const mediaWithMention: ReturnType<typeof generateMockPostMedia> = {
      ...generateMockPostMedia(),
      tappableObjects: [generateMockTappableObject()],
    };
    const mediaWithMentions: ReturnType<typeof generateMockPostMedia> = {
      ...generateMockPostMedia(),
      tappableObjects: [
        generateMockTappableObject({
          x: 0.1,
          y: 0.2,
          scale: 0.3,
          rotation: 40,
        }),
        generateMockTappableObject({
          x: 0.2,
          y: 0.3,
          scale: 0.4,
          rotation: 50,
        }),
      ],
    };

    it('renders mention tag on single-tap', async () => {
      const { UNSAFE_getByType, queryByTestId } = render(
        <PostItemImage media={mediaWithMention} />,
        options,
      );

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
      });

      expect(queryByTestId('MentionTag')).not.toBeNull();
    });

    it('hides mention tag on second single-tap', async () => {
      setupTimeTravel();
      const { UNSAFE_getByType, queryByTestId } = render(
        <PostItemImage media={mediaWithMention} />,
        options,
      );

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
      });

      expect(queryByTestId('MentionTag')).not.toBeNull();

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
        timeTravel(theme.animation.timingFast);
      });

      expect(queryByTestId('MentionTag')).toBeNull();
      destroyTimeTravel();
    });

    it('renders multiple mention tags on single-tap', async () => {
      const { UNSAFE_getByType, queryAllByTestId } = render(
        <PostItemImage media={mediaWithMentions} />,
        options,
      );

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
      });

      expect(queryAllByTestId('MentionTag')).toHaveLength(
        mediaWithMentions.tappableObjects.length,
      );
    });

    it('renders mention tags with provided config', async () => {
      const {
        UNSAFE_getByType,
        UNSAFE_getAllByType,
        queryAllByTestId,
      } = render(<PostItemImage media={mediaWithMentions} />, options);

      await act(async () => {
        UNSAFE_getByType(TapGestureHandler).props.onHandlerStateChange({
          nativeEvent: { state: State.ACTIVE },
        });
      });

      UNSAFE_getAllByType(Animated.View).forEach((animatedView, i) => {
        const tappable = mediaWithMentions.tappableObjects[i];

        expect(animatedView.props.style[0]).toEqual(
          expect.objectContaining({
            top: `${tappable.y * 100}%`,
            left: `${tappable.x * 100}%`,
          }),
        );
      });

      queryAllByTestId('MentionTag').forEach((tag, i) => {
        const tappable = mediaWithMentions.tappableObjects[i];

        expect(tag.props).toEqual(
          expect.objectContaining({
            rotation: tappable.rotation,
            scale: tappable.scale,
          }),
        );
        expect(tag).toHaveTextContent(tappable.text);
      });
    });
  });
});
