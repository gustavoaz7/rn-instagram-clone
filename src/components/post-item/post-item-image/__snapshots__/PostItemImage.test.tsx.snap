// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - PostItemImage matches snapshot 1`] = `
<View
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  testID="PostItemImage"
>
  <Image
    source={
      Object {
        "uri": "http://placeimg.com/640/480/cats?0",
      }
    }
    style={
      Array [
        Object {
          "height": 750,
          "width": 750,
        },
      ]
    }
  />
</View>
`;
